src/topology/locally_constant/basic.lean

```diff
@@ -0,0 +1,260 @@
+/-
+Copyright (c) 2021 Johan Commelin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johan Commelin
+-/
+import topology.subset_properties
+import topology.connected
+import topology.algebra.monoid
+
+/-!
+# Locally constant functions
+
+This file setups the theory of locally constant function from a topological space to a type.
+
+## Main definitions and constructions
+
+* `is_locally_constant f` : a map `f : X → Y` where `X` is a topological space is locally
+                            constant if everything set in `Y` has an open preimage.
+* `locally_constant X Y` : the type of locally constant maps from `X` to `Y`
+* `locally_constant.map` : push-forward of locally constant maps
+* `locally_constant.comap` : pull-back of locally constant maps
+
+-/
+
+variables {X Y Z α : Type*} [topological_space X]
+
+open_locale topological_space
+
+/-- A function between topological spaces is locally constant if the preimage of any set is open. -/
+def is_locally_constant (f : X → Y) : Prop := ∀ s : set Y, is_open (f ⁻¹' s)
+
+namespace is_locally_constant
+
+lemma exists_open {f : X → Y} (hf : is_locally_constant f) (x : X) :
+  ∃ (U : set X) (hU : is_open U) (hx : x ∈ U), ∀ x' ∈ U, f x' = f x :=
+⟨f ⁻¹' {(f x)}, hf _, set.mem_singleton _, λ x' hx', set.mem_singleton_iff.mp hx'⟩
+
+lemma exists_nhds {f : X → Y} (hf : is_locally_constant f) (x : X) :
+  ∃ U ∈ 𝓝 x, ∀ x' ∈ U, f x' = f x :=
+let ⟨U, hU, hx, H⟩ := hf.exists_open x in ⟨U, mem_nhds_sets hU hx, H⟩
+
+lemma iff_exists_open (f : X → Y) :
+  is_locally_constant f ↔ ∀ x, ∃ (U : set X) (hU : is_open U) (hx : x ∈ U), ∀ x' ∈ U, f x' = f x :=
+begin
+  refine ⟨exists_open, _⟩,
+  assume h s,
+  rw is_open_iff_forall_mem_open,
+  assume x hx,
+  obtain ⟨U, hU, hxU, H⟩ := h x,
+  refine ⟨U, _, hU, hxU⟩,
+  assume x' hx',
+  simp only [*, set.mem_preimage] at *,
+end
+
+lemma iff_exists_nhds (f : X → Y) :
+  is_locally_constant f ↔ ∀ x, ∃ U ∈ 𝓝 x, ∀ x' ∈ U, f x' = f x :=
+begin
+  refine ⟨exists_nhds, _⟩,
+  assume h,
+  rw iff_exists_open,
+  assume x,
+  obtain ⟨U, hU, H⟩ := h x,
+  obtain ⟨V, hVU, hV, hxV⟩ : ∃ (V : set X) (H : V ⊆ U), is_open V ∧ x ∈ V,
+  by rwa mem_nhds_sets_iff at hU,
+  refine ⟨V, hV, hxV, _⟩,
+  assume x' hx',
+  solve_by_elim only [H, hxV, hx', hVU]
+end
+
+lemma of_constant (f : X → Y) (h : ∃ y, ∀ x, f x = y) :
```

Comment:
Another possibility: `(h : ∀ x y, f x = f y)` (works for an empty `X`).

Reply:
Thanks, that's a nice improvement (-;